// ngx-hm-carousel/esm2015/lib/ngx-hm-carousel-dynamic.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Directive, Input, TemplateRef, ViewContainerRef } from '@angular/core';
export class NgxHmCarouselDynamicDirective {
    /**
     * @param {?} _view
     * @param {?} _template
     */
    constructor(_view, _template) {
        this._view = _view;
        this._template = _template;
        this.compelete = false;
        this._view.clear();
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set currentI(value) {
        if (!this.compelete) {
            const /** @type {?} */ nextI = value + 1;
            const /** @type {?} */ prevI = value - 1;
            if (this.index === 0 ||
                this.index === this.length - 1 ||
                this.index === nextI ||
                this.index === prevI ||
                this.index === value) {
                this._view.createEmbeddedView(this._template);
                this.compelete = true;
            }
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
}
NgxHmCarouselDynamicDirective.decorators = [
    { type: Directive, args: [{
                selector: '[ngxHmCarouselDynamic]'
            },] },
];
/** @nocollapse */
NgxHmCarouselDynamicDirective.ctorParameters = () => [
    { type: ViewContainerRef, },
    { type: TemplateRef, },
];
NgxHmCarouselDynamicDirective.propDecorators = {
    "index": [{ type: Input, args: ['ngxHmCarouselDynamic',] },],
    "length": [{ type: Input, args: ['ngxHmCarouselDynamicLength',] },],
    "currentI": [{ type: Input, args: ['ngxHmCarouselDynamicIndex',] },],
};
function NgxHmCarouselDynamicDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    NgxHmCarouselDynamicDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    NgxHmCarouselDynamicDirective.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    NgxHmCarouselDynamicDirective.propDecorators;
    /** @type {?} */
    NgxHmCarouselDynamicDirective.prototype.index;
    /** @type {?} */
    NgxHmCarouselDynamicDirective.prototype.length;
    /** @type {?} */
    NgxHmCarouselDynamicDirective.prototype.compelete;
    /** @type {?} */
    NgxHmCarouselDynamicDirective.prototype._view;
    /** @type {?} */
    NgxHmCarouselDynamicDirective.prototype._template;
}
